src/celengine/astrooctree.cpp

```diff
@@ -0,0 +1,389 @@
+{
+    if (m_children[n] == nullptr)
+        return false;
+    delete m_children[n];
+    m_children[n] = nullptr;
+    m_childrenCount--;
+    return true;
+}
+
+bool OctreeNode::insertObject(LuminousObject *obj)
+{
+    OctreeNode *node = this;
+//     cout << __LINE__ << "\n";
+    while (!node->isInCell(obj->getPosition()))
+    {
+        if (getParent() != nullptr)
+        {
+            node = node->getParent();
+        }
+        else
+            return false;
+    }
+//     cout << __LINE__ << "\n";
+    float mag = obj->getAbsoluteMagnitude();
+//     cout << __LINE__ << "\n";
+    while(node->getBrightest() > mag && node->getParent() != nullptr)
+    {
+        node = node->getParent();
+    }
+//     cout << __LINE__ << "\n";
+    while(node->getFaintest() < mag && node->m_objects.size() == node->m_maxObjCount)
+    {
+        node = node->getChild(node->getChildId(obj->getPosition()));
+    }
+//     cout << __LINE__ << "\n";
+    node->add(obj);
+//     cout << __LINE__ << "\n";
+    while(node->m_objects.size() > node->m_maxObjCount)
+    {
+        LuminousObject *obj = node->popFaintest();
+        node = node->getChild(node->getChildId(obj->getPosition()));
+        node->add(obj);
+    }
+//     cout << __LINE__ << "\n";
+    return true;
+}
+
+bool OctreeNode::removeObject(LuminousObject *obj)
+{
+    if (!rm(obj))
+    {
+        fmt::fprintf(cout, "Object %i not found to remove!\n", obj);
+        return false;
+    }
+    OctreeNode *node = this;
+    while(node->m_objects.size() < node->m_maxObjCount && node->getChildrenCount() > 0)
+    {
+//         fmt::fprintf(cout, "Node %i[%i/ %i]->%i\n", node, node->m_objects.size(), node->m_maxObjCount, node->getChildrenCount());
+        int i = node->getBrightestChildId();
+        if (i < 0)
+        {
+            fmt::fprintf(cout, "No brightest child of node %i!!\n", node);
+            getRoot()->dump(0);
+        }
+        OctreeNode *child = node->getChild(i);
+        if (child == nullptr)
+            cout << "Null child!!" << endl;
+        LuminousObject *obj = child->popBrightest();
+        node->add(obj);
+        if (child->getObjectCount() > 0)
+        {
+//             fmt::fprintf(cout, "Going to child %i.\n", child);
+            node = child;
+        }
+        else
+        {
+//             fmt::fprintf(cout, "Deleting child nr %i.\n", child);
+            node->deleteChild(i);
+        }
+    }
+    return true;
+}
+
+int OctreeNode::getChildId(const Eigen::Vector3d &pos)
+{
+    int child = 0;
+    child |= pos.x() < m_cellCenter.x() ? 0 : XPos;
+    child |= pos.y() < m_cellCenter.y() ? 0 : YPos;
+    child |= pos.z() < m_cellCenter.z() ? 0 : ZPos;
+
+    return child;
+}
+
+bool OctreeNode::isInFrustum(const Frustum::PlaneType *planes) const
+{
+    for (unsigned int i = 0; i < 5; ++i)
+    {
+        const Frustum::PlaneType& plane = planes[i];
+
+        float r = m_scale * plane.normal().cwiseAbs().sum() * 1.1;
```

Comment:
`* 1.1` looks like hack, we need to check Frustum calculation

Reply:
Indeed, however I spent already too many time and pain on it... For now I leave it for someone smarter.